### ISSUE TITLE
An issue with the accuracy of facenet

### ISSUE THREAD

**Issue by github4f** [davidsandberg/facenet#452]
Hi- I created my own pipeline based on mtcnn for face detection and then facenet for face recognition (I did test both 20170511-185253.pb and 20170512-110547.pb). I perform this process in a while loop for 2 minutes in which frames/images are captured using opencv video capture function! Finally, I compare my own images/faces during this period of time using cosine and euclidean functions! But the results are not promising at all. Indeed, the error can reach up to 50% sometime among frames! Any idea why is that and how I can improve it? 
Do I need to do normalization, or rgb2gray before giving my frames to mtcnn? What about alignment ( 3d alignment such as fb deep face)?

**Comment by fgervais** [comment]
I'm not sure if that would help or not but you can take a look at my pull request https://github.com/davidsandberg/facenet/pull/450. It seems to be right about what your are doing except for the distance measurement at the end.

It's mostly all about the classifier in the end but at least for me, with an SVM trained with my face among a hundred or so more faces from LFW, it can identify my face pretty much all the time at about 20 frames a seconds.

Sometimes the confidence level is not that high but it's always about 50% more than the next best guess.

**Comment by github4f** [comment]
Thanks for the note.

I was wondering how many images (your own face) was in the training set of SVM? 

I have two concerns with the classification: 1) when a new person is added to the Data Base, the whole SVM should be retrained! 2) In my application, it would be very hard to have access to more than couple of figures from the same person. How many figures I need from one person?

**Comment by fgervais** [comment]
I had about 20 images for each face.

The classifier is something I need to work a bit more on. I have a feeling that 20 images is too little but most faces in the LFW database have about 20 images. I didn't want to use too much pictures of my own face and having the classifier always guess the face as being me.

As for retraining the whole classifier, well retraining is not that bad depending on the hardware I guess. If not then maybe this?

http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html

**Comment by vudung45** [comment]
Check out my repo that I composed everything into a simple package, https://github.com/vudung45/FaceRec
The result is somewhat decent IMO. Fork or star it if you are interested.

**Comment by github4f** [comment]
Hi- I did an experiment using about 120 figures of myself. You can find the hist in the following figure. According to the results, normally I see more than 75% similarity between my pictures, but sometime it drops below 70%!

<img width="360" alt="myface" src="https://user-images.githubusercontent.com/6980446/30273993-c66b40d0-96fb-11e7-8338-baaaa32cd141.PNG">

In the following figure, I compare my own figures (120 figures) vs 120 figures of a friend of mine:

<img width="359" alt="i_vs_my friend" src="https://user-images.githubusercontent.com/6980446/30274091-33d2bb30-96fc-11e7-883a-b4f42377001a.PNG">

It is very interesting that sometime the similarity between I and my friend is beyond 60%. Based on the above figures, I decided to use 75% as the threshold in my code. 

@vudung45 : I am wondering your code is consistent with my results or you reach better similarity among your figures (for just one person)? I noticed that you use a different approach for alignment. 

I also did some investigation to see why there are so much similarity between I and my friend. I noticed that most of the time, the problem is rooted in wrong captured figures. for example, there are some figures that both eyes are not visible in the figures. Something like below: 

![figure drawing for dummies -mantesh-80-1](https://user-images.githubusercontent.com/6980446/30274213-b9d717b2-96fc-11e7-8a21-e58b224eea4b.jpg)

I noticed that openface (http://cmusatyalab.github.io/openface/), does use opencv cascade (eye), to detect false figures. But mtcnn cannot do that! I suggest that we also include eye cascade to remove those figures. Please let me know your idea (@fgervais ).

**Comment by github4f** [comment]
@vudung45 : Hi- I did test your code. It seems that the accuracy is not higher than the accuracy of davidsandberg model, although you have 3 different alignment metrics (left, center, right) in your database for comparsion! Did you train the facenet again? Do you have any new checkpoint for yourself? I noted that you just published a small version of your code. Maybe that is the reason we see a lower accuracy (e.g., when I move my head a little bit (30 degree), your code cannot recognize me). I appreciate your comment.

**Comment by vudung45** [comment]
@github4f 
First, Thank you for checking it out. Here are my 2 cents that could possible solve your accuracy problem.
1. If you want to test out my code more, try inputting new inputing like this:
![GIF Demo](https://media.giphy.com/media/3o7aD7CZ6C3RLCvLgs/giphy.gif)

2. Given the constrain of the facenet model's accuracy, there are many ways you can improve the accuracy in real world application. One of my suggestion would be to create a tracker for each detected face on screen, then run recognition on each of them in real time. Then, decide who is in each tracker after some number of frames ( 3 - 10 frames, depending on how fast your machine is). Keep doing the same thing until the tracker disappears or loses track. Your result can look somewhat like this: 
`{"Unknown" :3, "PersonA": 1, "PersonB": 20`}` ---> This tracker is tracking PersonB
 This will definitely help with your accuracy problem, because the result will most likely leaning toward the right subject in the picture after some number of frames instead of one.

3. My algorithm to find right , center, left is pretty simple. I decide it based on MTCNN's facial landmarks output

**Comment by github4f** [comment]
@vudung45 Hi David- Thanks a lot for the explanation and your great ideas concerning the matter we discussed. 

-I tried to use tracking techniques before to reduce the runtime issue which I have faced. Here is my question in stackoverlfow regarding this issue:
 https://stackoverflow.com/questions/46091964/boost-face-detection-recognition-process-in-python-opencv-tensorflow-cnn

-I was wondering if you have already implemented the tracking idea in your bigger project? Do you use opencv for the implementation e.g., MIL, KCF? 

-The issue with tracking (in particular the implementation with opencv) is that sometime (e.g., when the user moves a little fast) it shows false result (false rect)! At least opencv suffers form this issue. That is why, I prefer to use face detection and recognition for each frame rather than using face detection/recognition in 1 frame and tracking for the upcoming frames to avoid the issue. I appreciate it if you could kindly share your experience with us.

-Regarding  "inputting new inputing" which you stated: I am wondering if maximum 3 embedding (left, right, center) can be saved in the database for each person. Am I correct? or is it possible for example, I save 2 embedding for right, 2 for center, and 2 for left for just 1 person? 

Thank you

**Comment by github4f** [comment]
@vudung45  One more issue regarding the code. Now you calculate the similarity percentage using the following equation: percentage =  min(100, 100 * thres / smallest), where thre=0.6!  I am note the equation is correct or not, since it depends on your own thres! I prefer to use cosine similarity which gives a value between 0 and 1. So you can directly convert it to percentage without requiring thres.

**Comment by github4f** [comment]
@vudung45 
It seems that the following lines are also different:

`
 rects, landmarks = face_detect.detect_face(frame, 80);  # min face size is set to 80x80
aligned_frame, pos = aligner.align(160,frame,landmarks[i]);
`

According to mtcc the output should be like the following:
<img width="65" alt="correct" src="https://user-images.githubusercontent.com/6980446/30329144-c23e1e16-97d1-11e7-8183-75a393ccc65e.PNG">

But when I run your code, I see the following output (below the lips is ignored):

<img width="65" alt="mine" src="https://user-images.githubusercontent.com/6980446/30329164-cfc390fc-97d1-11e7-9ba0-2a05981be467.PNG">

And when I compare it with you comment (your own figure which you sent it here), I see the following:

<img width="65" alt="yours" src="https://user-images.githubusercontent.com/6980446/30329206-efe50cf8-97d1-11e7-8b11-f5661fbd1535.PNG">

Upper your eyebrow is not in the figure!

-So which one is correct?
-Why you use landmark to detect the face? I use directly "rects". I think rects has the whole face. but in you do some process on landmark and cut some part of the face. Am I correct? Do you have another code for alignment?

**Comment by github4f** [comment]
@vudung45  I also observed that your function does not work correctly:
`
rects, landmarks = face_detect.detect_face(image_list[0], 80);  # min face size is set to 80x80
`
When I use "rects" to draw a rect around the face, it shows something very irreverent. But when I use the official alignment of davidsandberg, it works fine. I suggest that you load an image:

`
im=misc.imread(filename)
`
and apply your facedetection on the image and drat the rect around the face to verify your function. If you have an updated function, please update the github so we can use it. Thank you

**Comment by vudung45** [comment]
@github4f 
Sorry for lack of comments, I was just breaking it from my other project.
1. My mtcnn detection is not entirely the same as davidsandberg's one. It changed its datastucture up a little bit for my own comfort. If I remember correctly, each rect in rects is something like this:
(x,y,w,h)
2. Rect is indeed the whole face, however, my AlignCustom class uses the 5 landmarks to align the face and try to crop it in a way that inner eyes, nose, and outer lips are around the "average positions of face points".
3. I have implemented the tracking method in my other project. OpenCV trackers worked pretty decently. I used them mostly to get predicted updated locations of the faces in the next frame then match them with the faces on screen then "update" the width and height of the bounding boxes since OpenCV trackers don't do that for you. 
4. For inputting new subject matter, I find the mean/center of each of the 3 face position(left, right, center). By doing that, the more data you have, the better center point is.

**Comment by github4f** [comment]
@vudung45  Thanks a lot for your comments. 
-Ya, I exactly use the rect output of your mtcnn based on (x,y,w,h). But it does not give correct output. But the landmark is fine and great. Do you have an updated version of your mtcnn?

-I am also wondering if you have any repository for "tracker". As you mentioned, opencv does no work. It would be great, if you could share it with us. 

Thanks

**Comment by vudung45** [comment]
@github4f 
- Is the output slightly off? That might have been because of the padding parameter I create for the align function, which is set to 0.1 as default. Try to adjust that if you want its to crop a bigger face region.
- As for tracker, I did make my own tracker class implementing/reinforcing OpenCV tracker. I think I will be putting it together with this repo later. 

Once again, thank you for checking it out :)

**Comment by github4f** [comment]
@vudung45  Not really, the returned rect is so wrong! See the following figure which is the output of your code: 

<img width="167" alt="wrong-rect" src="https://user-images.githubusercontent.com/6980446/30375907-98354914-988a-11e7-9ba9-1ea5a04be54d.png">


btw, what kind of mtcnn you use? For me, sometime, it detects false faces like the below figure:

<img width="408" alt="fault" src="https://user-images.githubusercontent.com/6980446/30375939-b61da296-988a-11e7-87bb-2987d5a52752.PNG">

**Comment by cjs210** [comment]
I had the same issue, and i have been trying to solve it. Here comes my two cents:

1) Every facial recognition algorithm will be affected by changes in illumination, if you want to make your system less sensitive, your training images should have different light settings, so that the result of recognition doesn't drop to drastically.

2) I have reached 85% of accuracy (90% a few times) of facial recognition using Haar Cascade available on OpenCV for face detection. The Haar Cascade file used was lbpcascade_frontalface_improved.xml if you want to try, here is the link -> https://github.com/opencv/opencv/blob/master/data/lbpcascades/lbpcascade_frontalface_improved.xml (No, i am not using LBP for face detection, the file was made using LBP, but you still have to use Haar Cascade)

It may not be as accurate as using MTCNN but it has been accurate enough for me and you can change settings to improve accuracy if you have to.

You have to supply training data using Haar Cascade too. Make an algorithm to detect faces using Haar Cascade, save the images, train your system with those faces and voila, accuracy improved! At least that was my experience using Facenet.
